pkg/pool-linear/contracts/reaper/ReaperLinearPool.sol

```diff
@@ -0,0 +1,95 @@
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+pragma solidity ^0.7.0;
+pragma experimental ABIEncoderV2;
+
+import "@balancer-labs/v2-interfaces/contracts/pool-linear/IReaperTokenVault.sol";
+
+import "../LinearPool.sol";
+
+contract ReaperLinearPool is LinearPool {
+    IReaperTokenVault private immutable _tokenVault;
+
+    struct ConstructorArgs {
+        IVault vault;
+        string name;
+        string symbol;
+        IERC20 mainToken;
+        IERC20 wrappedToken;
+        address assetManager;
+        uint256 upperTarget;
+        uint256 swapFeePercentage;
+        uint256 pauseWindowDuration;
+        uint256 bufferPeriodDuration;
+        address owner;
+    }
+
+    constructor(ConstructorArgs memory args)
+        LinearPool(
+            args.vault,
+            args.name,
+            args.symbol,
+            args.mainToken,
+            args.wrappedToken,
+            args.upperTarget,
+            _toAssetManagerArray(args),
+            args.swapFeePercentage,
+            args.pauseWindowDuration,
+            args.bufferPeriodDuration,
+            args.owner
+        )
+    {
+        IReaperTokenVault tokenVault = IReaperTokenVault(address(args.wrappedToken));
+        
+        _tokenVault = tokenVault;
+        _require(address(args.mainToken) == tokenVault.token(), Errors.TOKENS_MISMATCH);
+    }
+
+    function _toAssetManagerArray(ConstructorArgs memory args) private pure returns (address[] memory) {
+        // We assign the same asset manager to both the main and wrapped tokens.
+        address[] memory assetManagers = new address[](2);
+        assetManagers[0] = args.assetManager;
+        assetManagers[1] = args.assetManager;
+
+        return assetManagers;
+    }
+
+    /**
+     * Reaper vault tokens are always 18 decimal places, but are stored in the precision of the mainToken
+     * ie: If ppfs === 1e18, 1 USDC === 0.000_000_000_001_000_000 rfUSDC
+     * ie: If ppfs === 1e18, 1 DAI === 1 rfDAI
+     * ie: If ppfs === 1e18, 1 WBTC === 0.000_000_000_100_000_000 rfWBTC
+     * -----------
+     * Internally, the LinearPool scales all balances and rates up to 18 decimal places, meaning that 1 USDC is 
+     * represented as 1e18 by the LinearPool. Since the rfUSDC is already 18 decimals,
+     * but in a different representation,we need to account for that in our wrappedTokenRate.
+     */
+    function _getWrappedTokenRate() internal view override returns (uint256) {
+        ERC20 underlying = ERC20(_tokenVault.token());
+        uint256 underlyingDecimals = underlying.decimals();
+
+        uint256 sharePrice = _tokenVault.getPricePerFullShare();
```

Comment:
I'm going to assume this meets all the LinearPool rate requirements: critically, this value must not be manipulable, and anyone should be able to get the exact same rate by interacting with Reaper.

Reply:
![2022-09-16 11 04 58](https://user-images.githubusercontent.com/91405705/190600861-773a9018-5768-4b92-8ac8-565adb2d0287.jpg)

Attached is an excerpt from a conversation with a yearn dev. The same applies for reaper vaults. It's possible to manipulate the ppfs upwards, but the cost of this attack is not financially viable for the linear pool. All reaper vaults we will deploy linear pools for will be non upgradable single strategy vaults where assets are deposited into AAVE (no folding).

Reply:
I'd argue that'd be a true price change, and not manipulation (since they cannot undo the change). That sort of behavior is fine.

The other important requirement is that the rate accurately reflects the underlying exchange rate. The LinearPool offers trades at that exact rate with no swap fees of any kind: if the rate is off even by a small percentage, then the Pool can be trivially drained by swapping with it back and forth.

Reply:
Price per full share accurately represents the current and real time exchange rate of vault token to underlying asset 👍🏽